birdhouse/components/weaver/config/magpie/weaver_hooks.py.template

```diff
@@ -47,20 +64,16 @@ def add_x_wps_output_context(request):
     Apply the ``X-WPS-Output-Context`` for saving outputs in the user-context WPS-outputs directory.
     """
     header = get_header("X-WPS-Output-Context", request.headers)
-    # if explicitly provided, ensure it is permitted (admin allow any, otherwise self-user reference only)
-    if header is not None:
-        if request.user is None:
-            header = "public"
-        else:
-            if not is_admin(request):
-                # override disallowed writing to other location
-                # otherwise, up to admin to have written something sensible
-                header = "users/" + str(request.user.id)
-    else:
-        if request.user is None:
-            header = "public"
-        else:
-            header = "users/" + str(request.user.id)
+    # if no user logged in, ignore specified value to avoid writing in someone else's user-workspace
+    # if explicitly requested to make their job output public right away, whether logged in or not,
+    # use the public location instead of specific user-workspace
+    if request.user is None or header in PUBLIC_USER_CONTEXTS:
+        header = PUBLIC_WORKSPACE_WPS_OUTPUTS_SUBDIR
+    # if explicitly provided, ensure it is permitted, and if not provided, default to current user
+    # admin-level user is allowed to write anywhere (up to them to use a sensible location)
+    # otherwise, enforce the current user-workspace even if provided explicitly
+    if not header or not is_admin(request):
```

Comment:
If `request.user is None` then `is_admin(request)` is always going to return False right?

So, in this case wouldn't the next line raise an error since `request.user` is `None` and so `request.user.id` raises an AttributeError?

If we make this line an `elif` instead of a new `if` it should work though since the case where `request.user is None` is handled on line 70.

Reply:
Good catch. I guess that's why I did the `if request.user is None:` on its own in the original implementation.

Reply:
I'll just revert to something separate and stop wasting time on refactoring these `if` combinations.

Reply:
I like the code better now. It's much easier to read :+1: